exporter/otelarrowexporter/metadata.go

```diff
@@ -0,0 +1,201 @@
+	e.exporters.Range(func(_ any, value any) bool {
+		be, ok := value.(exp)
+		if !ok {
+			err = multierr.Append(err, fmt.Errorf("%w: %T", errUnexpectedType, value))
+			return true
+		}
+		err = multierr.Append(err, be.shutdown(ctx))
+		return true
+	})
+	return err
+}
+
+func (e *metadataExporter) pushTraces(ctx context.Context, td ptrace.Traces) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+	return be.pushTraces(ctx, td)
+}
+
+func (e *metadataExporter) pushMetrics(ctx context.Context, md pmetric.Metrics) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+
+	return be.pushMetrics(ctx, md)
+}
+
+func (e *metadataExporter) pushLogs(ctx context.Context, ld plog.Logs) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+
+	return be.pushLogs(ctx, ld)
+}
+
+func (e *metadataExporter) getOrCreateExporter(ctx context.Context, s attribute.Set, md metadata.MD) (exp, error) {
+	v, ok := e.exporters.Load(s)
+	if !ok {
+		e.lock.Lock()
+		if e.config.MetadataCardinalityLimit != 0 && e.size >= int(e.config.MetadataCardinalityLimit) {
+			e.lock.Unlock()
+			return nil, errTooManyExporters
+		}
+
+		newExp, err := newExporter(e.config, e.settings, e.scf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create exporter: %w", err)
+		}
+
+		var loaded bool
+		v, loaded = e.exporters.LoadOrStore(s, newExp)
+		if !loaded {
+			// Start the goroutine only if we added the object to the map, otherwise is already started.
+			be, valid := newExp.(*baseExporter)
+			if !valid {
+				return nil, fmt.Errorf("%w: %T", errUnexpectedType, newExp)
+			}
+			// set metadata keys for base exporter to add them to the outgoing context.
+			be.setMetadata(md)
+
+			err = newExp.start(ctx, e.host)
+			if err != nil {
+				e.exporters.Delete(s)
+				return nil, fmt.Errorf("failed to start exporter: %w", err)
+			}
+			e.size++
+		}
+		e.lock.Unlock()
+	}
+	val, ok := v.(exp)
+	if !ok {
+		return nil, fmt.Errorf("%w: %T", errUnexpectedType, v)
+	}
+	return val, nil
+}
+
+func (e *metadataExporter) getAttrSet(ctx context.Context, keys []string) (attribute.Set, metadata.MD) {
+	// Get each metadata key value, form the corresponding
+	// attribute set for use as a map lookup key.
+	info := client.FromContext(ctx)
+	md := map[string][]string{}
+	var attrs []attribute.KeyValue
+	for _, k := range keys {
+		// Lookup the value in the incoming metadata, copy it
+		// into the outgoing metadata, and create a unique
+		// value for the attributeSet.
+		vs := info.Metadata.Get(k)
+		md[k] = vs
+		if len(vs) == 1 {
+			attrs = append(attrs, attribute.String(k, vs[0]))
+		} else {
+			attrs = append(attrs, attribute.StringSlice(k, vs))
+		}
+	}
+	return attribute.NewSet(attrs...), metadata.MD(md)
+}
```

Comment:
:+1: Maybe a note saying this is copied from https://github.com/open-telemetry/opentelemetry-collector/blob/b33913ba527d8ca9ddac79c82760b61aa6192b1b/processor/batchprocessor/batch_processor.go#L296C1-L299C1 for the record.

We may as well remember https://github.com/open-telemetry/opentelemetry-collector/issues/7455 here. :-)

Reply:
Added a comment. Not sure I understand your point about https://github.com/open-telemetry/opentelemetry-collector/issues/7455